src/Utilities/StdConstexprHelpers.hpp

```diff
@@ -0,0 +1,116 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/*!
+ * \ingroup Utilities
+ * These functions are constexpr reimplementations of some standard library
+ * functions, designed to work with cpp17::array.
+ */
+
+#pragma once
+
+#include "Utilities/Gsl.hpp"
+
+namespace cpp2b {
+/*!
+ * \ingroup Utilities
+ * Reimplementation of std::accumulate that is constexpr;
+ * taken from the LLVM source at
+ * https://github.com/llvm-mirror/libcxx/blob/master/include/numeric
+ */
+template <class ForwardIterator, class T>
+constexpr void iota(ForwardIterator first, ForwardIterator last,
+                    T value) noexcept {
+  for (; first != last; ++first, (void)++value) {
+    *first = value;
+  }
+}
+
+/*!
+ * \ingroup Utilities
+ * Reimplementation of std::accumulate that is constexpr;
+ * taken from the LLVM source at
+ * https://github.com/llvm-mirror/libcxx/blob/master/include/numeric
+ */
+template <class InputIt, class T>
+constexpr T accumulate(InputIt first, InputIt last, T init) noexcept {
+  for (; first != last; ++first) {
+    init = std::move(init) + *first;  // std::move since C++20
+  }
+  return init;
+}
+}  // namespace cpp2b
+
+namespace cpp20 {
+/*!
+ * \ingroup Utilities
+ * Reimplementation of std::swap that is constexpr;
+ * taken from the LLVM source at
+ * https://github.com/llvm-mirror/libcxx/blob/master/include/type_traits
+ */
+template <class T>
+constexpr void swap(T& a, T& b) noexcept {
+  T c(std::move(a));
+  a = std::move(b);
+  b = std::move(c);
+}
+
+/*!
+ * \ingroup Utilities
+ * Reimplementation of std::iter_swap that is constexpr;
+ * taken from the LLVM source at
+ * https://github.com/llvm-mirror/libcxx/blob/master/include/type_traits
+ */
+template <class ForwardIt1, class ForwardIt2>
+constexpr void iter_swap(ForwardIt1 a, ForwardIt2 b) noexcept {
+  swap(*a, *b);
+}
+
+/*!
+ * \ingroup Utilities
+ * Reimplementation of std::reverse that is constexpr;
+ * taken from the LLVM source at
+ * https://github.com/llvm-mirror/libcxx/blob/master/include/algorithm
+ */
+template <class BidirIt>
+constexpr void reverse(BidirIt first, BidirIt last) noexcept {
```

Comment:
This is only the implementation for bidirectional iterators, not the random access ones.